filters/include/pcl/filters/experimental/voxel_grid.h

```diff
@@ -0,0 +1,509 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2020-, Open Perception
+ *
+ *  All rights reserved
+ */
+
+#pragma once
+
+#include <pcl/common/centroid.h>
+#include <pcl/common/common.h>
+#include <pcl/filters/experimental/cartesian_filter.h>
+
+#include <unordered_map>
+
+namespace pcl {
+namespace experimental {
+
+/**
+ * \brief VoxelStructT defines the transformation operations and the voxel grid of
+ * VoxelGrid filter \ingroup filters
+ */
+template <typename PointT>
+class VoxelStructT {
+public:
+  using PointCloud =
+      pcl::PointCloud<PointT>; // read by CartesianFilter to deduce point type
+  using PointCloudPtr = typename PointCloud::Ptr;
+  using PointCloudConstPtr = typename PointCloud::ConstPtr;
+  using Grid = typename std::unordered_map<std::size_t, CentroidPoint<PointT>>;
+  using GridIterator = typename Grid::iterator;
+
+  /** \brief Empty constructor. */
+  VoxelStructT() { filter_name_ = "VoxelGrid"; }
+
+  /** \brief Get the number of voxels in the grid. */
+  std::size_t
+  size() const
+  {
+    return grid_.size();
+  }
+
+  /** \brief Get the begin iterator of the grid. */
+  GridIterator
+  begin()
+  {
+    return grid_.begin();
+  }
+
+  /** \brief Get the end iterator of the grid. */
+  GridIterator
+  end()
+  {
+    return grid_.end();
+  }
+
+  /** \brief Set up the voxel grid variables needed for filtering.
+   * \param[in] transform_filter pointer to TransformFilter
+   */
+  bool
+  setUp(TransformFilter<VoxelStructT>* transform_filter)
+  {
+    num_voxels_ = 0;
+    grid_ = Grid();
```

Comment:
This is the part having the most runtime increase. It used to be `grid_.clear()`, but I found this is faster than `clear()` after some trials

Reply:
http://en.cppreference.com/w/cpp/container/unordered_map/clear

Complexity of clear is linear in number of elements.

Funnily, the destructor and the ctor have the same behavior. Could you use perf to dig into the issue?

Reply:
FYI, there are basically no collisions in the `unordered_map`, only 1 collision across all benchmarked clouds.

And the load_factor looks normal too

| cloud   | load_factor |
|---------|-------------|
| 5people | 0.356992    |
| office  | 0.557616    |
| milk    | 0.0778205   |
| mug     | 0.0289335   |

Reply:
0.55 is quite large. What happens if we call `unordered_map::max_load_factor(X);` where X is some number from 0.4 to 0.5?

Reply:
Benchmarked with `for_each` & `clear`

### without `max_load_factor`
```
BM_VoxelGrid_5people                    13.3 ms         13.3 ms           55
BM_Experimental_VoxelGrid_5people       10.0 ms         10.0 ms           69
BM_VoxelGrid_office                     13.6 ms         13.6 ms           52
BM_Experimental_VoxelGrid_office        14.1 ms         14.1 ms           48
```

### `max_load_factor(0.5)`
```
BM_VoxelGrid_5people                    13.5 ms         13.5 ms           53
BM_Experimental_VoxelGrid_5people       9.18 ms         9.18 ms           77
BM_VoxelGrid_office                     13.6 ms         13.6 ms           52
BM_Experimental_VoxelGrid_office        12.2 ms         12.2 ms           55
```

### `max_load_factor(0.4)`
```
BM_VoxelGrid_5people                    13.1 ms         13.1 ms           53
BM_Experimental_VoxelGrid_5people       8.74 ms         8.74 ms           75
BM_VoxelGrid_office                     12.9 ms         12.9 ms           54
BM_Experimental_VoxelGrid_office        11.3 ms         11.3 ms           59
```

Reply:
A minor question: for reserving the capacity of a hash-map, which method are you using:
* `hash_map.reserve(capacity);`
* `hash_map.reserve(std::ceil(capacity / has_map.max_load_factor()));`

If the former, then you are also hurting performance by causing some rehashing once the load factor is reached. If so, please use the latter method, and re-run the benchmarks

Reply:
> Funnily, the destructor and the ctor have the same behavior. Could you use perf to dig into the issue?

It is indeed because of constructing and destructing: `_int_malloc` and  `_int_free` occupied considerable amount of runtime with `unordered_map::clear()`

But if we look at the original `VoxelGrid`, it also creates the same number of `CentroidPoint`. I guess that part is optimized away by the compiler? But I'm not an expert of that
https://github.com/PointCloudLibrary/pcl/blob/97b0f0e63b3325bf8a19c99d0f0918b9cd86eb75/filters/include/pcl/filters/impl/voxel_grid.hpp#L393-L423

## `perf` results

### `unordered_map::clear()`
```
    43.46%  benchmark_filte  benchmark_filters_voxel_grid  [.] pcl::experimental::TransformFilter<pcl::experimental::VoxelStructT<pcl::PointXYZ>, pcl::PointXYZ>::applyFilter
    13.10%  benchmark_filte  libc-2.31.so                  [.] _int_malloc
    10.11%  benchmark_filte  libc-2.31.so                  [.] _int_free
     6.16%  benchmark_filte  benchmark_filters_voxel_grid  [.] std::_Hashtable<unsigned long, std::pair<unsigned long const, pcl::experimental::Voxel<pcl::PointXYZ> >, std::allocator<std::pair<unsigned long const, pcl::experimental::Voxel<pcl::PointXYZ> > >, std::__detail::_Select1st, std::equal_to<unsigned long>, std::hash<unsigned long>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::_M_emplace<unsigned long const&, bool&>
     4.43%  benchmark_filte  benchmark_filters_voxel_grid  [.] pcl::experimental::VoxelStructT<pcl::PointXYZ>::setUp
     3.67%  benchmark_filte  libc-2.31.so                  [.] __mcount_internal
     3.31%  benchmark_filte  libc-2.31.so                  [.] malloc
     1.70%  benchmark_filte  libc-2.31.so                  [.] malloc_consolidate
     1.58%  benchmark_filte  libpcl_io.so.1.12.0.99        [.] pcl::lzfDecompress
     1.38%  benchmark_filte  libc-2.31.so                  [.] cfree@GLIBC_2.2.5
     1.25%  benchmark_filte  libc-2.31.so                  [.] _mcount
```

### `for_each + clear()`
```
68.74%  benchmark_filte  benchmark_filters_voxel_grid   [.] pcl::experimental::TransformFilter<pcl::experimental::VoxelStructT<pcl::PointXYZ>, pcl::PointXYZ>::applyFilter
    16.51%  benchmark_filte  benchmark_filters_voxel_grid   [.] pcl::experimental::VoxelStructT<pcl::PointXYZ>::setUp
     2.00%  benchmark_filte  libpcl_io.so.1.12.0.99         [.] pcl::lzfDecompress
     1.42%  benchmark_filte  libpcl_io.so.1.12.0.99         [.] pcl::PCDReader::readBodyBinary
     0.82%  benchmark_filte  libc-2.31.so                   [.] malloc_consolidate
     0.82%  benchmark_filte  benchmark_filters_voxel_grid   [.] std::_Hashtable<unsigned long, std::pair<unsigned long const, pcl::experimental::Voxel<pcl::PointXYZ> >, std::allocator<std::pair<unsigned long const, pcl::experimental::Voxel<pcl::PointXYZ> > >, std::__detail::_Select1st, std::equal_to<unsigned long>, std::hash<unsigned long>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::_M_emplace<unsigned long const&, bool&>
     0.77%  benchmark_filte  libc-2.31.so                   [.] __memmove_avx_unaligned_erms
     0.69%  benchmark_filte  ld-2.31.so                     [.] do_lookup_x
     0.65%  benchmark_filte  libc-2.31.so                   [.] _int_free
     0.65%  benchmark_filte  libc-2.31.so                   [.] _int_malloc
     0.63%  benchmark_filte  libc-2.31.so                   [.] __mcount_internal
     0.53%  benchmark_filte  libc-2.31.so                   [.] malloc
     0.38%  benchmark_filte  libc-2.31.so                   [.] unlink_chunk.isra.0
```

### original `VoxelGrid`
```
  35.00%  benchmark_filte  libpcl_filters.so.1.12.0.99   [.] pcl::VoxelGrid<pcl::PointXYZ>::applyFilter
  20.47%  benchmark_filte  libpcl_filters.so.1.12.0.99   [.] boost::sort::pdqsort_detail::pdqsort_loop<__gnu_cxx::__normal_iterator<cloud_point_index_idx*, std::vector<cloud_point_index_idx, std::allocator<cloud_point_index_idx> > >, std::less<cloud_point_index_idx>, false>
  13.31%  benchmark_filte  libpcl_filters.so.1.12.0.99   [.] boost::sort::spreadsort::detail::spreadsort_rec<__gnu_cxx::__normal_iterator<cloud_point_index_idx*, std::vector<cloud_point_index_idx, std::allocator<cloud_point_index_idx> > >, unsigned int, pcl::VoxelGrid<pcl::PointXYZ>::applyFilter(pcl::PointCloud<pcl::PointXYZ>&)::{lambda(clo
   5.07%  benchmark_filte  libc-2.31.so                  [.] _mcount
   4.02%  benchmark_filte  libc-2.31.so                  [.] __mcount_internal
   3.62%  benchmark_filte  libpcl_filters.so.1.12.0.99   [.] std::vector<cloud_point_index_idx, std::allocator<cloud_point_index_idx> >::emplace_back<unsigned int, int const&>
   2.89%  benchmark_filte  libpcl_filters.so.1.12.0.99   [.] std::vector<std::pair<unsigned int, unsigned int>, std::allocator<std::pair<unsigned int, unsigned int> > >::emplace_back<unsigned int&, unsigned int&>
   2.00%  benchmark_filte  libpcl_io.so.1.12.0.99        [.] pcl::lzfDecompress
   1.51%  benchmark_filte  ld-2.31.so                    [.] do_lookup_x
   1.47%  benchmark_filte  libpcl_io.so.1.12.0.99        [.] pcl::PCDReader::readBodyBinary
   1.29%  benchmark_filte  libpcl_filters.so.1.12.0.99   [.] boost::sort::spreadsort::detail::is_sorted_or_find_extremes<__gnu_cxx::__normal_iterator<cloud_point_index_idx*, std::vector<cloud_point_index_idx, std::allocator<cloud_point_index_idx> > > >
   0.94%  benchmark_filte  libpcl_filters.so.1.12.0.99   [.] 0x0000000000331d44
   0.80%  benchmark_filte  libc-2.31.so                  [.] __memmove_avx_unaligned_erms
   0.77%  benchmark_filte  ld-2.31.so                    [.] _dl_lookup_symbol_x
```

Reply:
I did see some improvements in 2 runs where we edit the max_load_factor, but not where we're using the default. Could you please confirm if the 1st run was also reserved considering load factor in mind?

PS: Looking at the perf results soon

Reply:
> Could you please confirm if the 1st run was also reserved considering load factor in mind?

I benchmark `unordered_map::clear()` as contrast to `for_each` & `clear` above

### `without max_load_factor`
`BM_Experimental_VoxelGrid_office       45.8 ms         45.8 ms           17`

```
    45.13%  benchmark_filte  benchmark_filters_voxel_grid  [.] pcl::experimental::TransformFilter<pcl::experimental::VoxelStructT<pcl::PointXYZ>, pcl::PointXYZ>::applyFilter
    13.72%  benchmark_filte  libc-2.31.so                  [.] _int_malloc
    10.29%  benchmark_filte  libc-2.31.so                  [.] _int_free
     5.02%  benchmark_filte  benchmark_filters_voxel_grid  [.] std::_Hashtable<unsigned long, std::pair<unsigned long const, pcl::experimental::Voxel<pcl::PointXYZ> >, std::allocator<std::pair<unsigned long const, pcl::experimental::Voxel<pcl::PointXYZ> > >, std::__detail::_Select1st, std::equal_to<unsigned long>, std::hash<unsigned long>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::_M_emplace<unsigned long const&, bool&>
     4.98%  benchmark_filte  benchmark_filters_voxel_grid  [.] pcl::experimental::VoxelStructT<pcl::PointXYZ>::setUp
     3.34%  benchmark_filte  libc-2.31.so                  [.] malloc
     3.25%  benchmark_filte  libc-2.31.so                  [.] __mcount_internal
     1.73%  benchmark_filte  libc-2.31.so                  [.] malloc_consolidate
     1.34%  benchmark_filte  libc-2.31.so                  [.] cfree@GLIBC_2.2.5
     1.33%  benchmark_filte  libpcl_io.so.1.12.0.99        [.] pcl::lzfDecompress
     1.02%  benchmark_filte  libc-2.31.so                  [.] _mcount
     1.00%  benchmark_filte  libpcl_io.so.1.12.0.99        [.] pcl::PCDReader::readBodyBinary
     0.98%  benchmark_filte  libstdc++.so.6.0.28           [.] operator new
```

### `max_load_factor(0.5)`
`BM_Experimental_VoxelGrid_office       38.5 ms         38.5 ms           18`

```
    41.09%  benchmark_filte  benchmark_filters_voxel_grid       [.] pcl::experimental::TransformFilter<pcl::experimental::VoxelStructT<pcl::PointXYZ>, pcl::PointXYZ>::applyFilter
    13.74%  benchmark_filte  libc-2.31.so                       [.] _int_malloc
     9.42%  benchmark_filte  libc-2.31.so                       [.] _int_free
     5.98%  benchmark_filte  benchmark_filters_voxel_grid       [.] std::_Hashtable<unsigned long, std::pair<unsigned long const, pcl::experimental::Voxel<pcl::PointXYZ> >, std::allocator<std::pair<unsigned long const, pcl::experimental::Voxel<pcl::PointXYZ> > >, std::__detail::_Select1st, std::equal_to<unsigned long>, std::hash<unsigned long>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::_M_emplace<unsigned long const&, bool&>
     4.81%  benchmark_filte  benchmark_filters_voxel_grid       [.] pcl::experimental::VoxelStructT<pcl::PointXYZ>::setUp
     4.05%  benchmark_filte  libc-2.31.so                       [.] malloc
     4.05%  benchmark_filte  libc-2.31.so                       [.] __mcount_internal
     1.66%  benchmark_filte  libc-2.31.so                       [.] cfree@GLIBC_2.2.5
     1.58%  benchmark_filte  libpcl_io.so.1.12.0.99             [.] pcl::lzfDecompress
     1.46%  benchmark_filte  libc-2.31.so                       [.] _mcount
     1.16%  benchmark_filte  libpcl_io.so.1.12.0.99             [.] pcl::PCDReader::readBodyBinary
     1.14%  benchmark_filte  libc-2.31.so                       [.] malloc_consolidate
     0.99%  benchmark_filte  libc-2.31.so                       [.] __memset_avx2_erms
     0.93%  benchmark_filte  libstdc++.so.6.0.28                [.] operator new
```

### `max_load_factor(0.4)`
`BM_Experimental_VoxelGrid_office       32.0 ms         32.0 ms           23`

```
    40.15%  benchmark_filte  benchmark_filters_voxel_grid  [.] pcl::experimental::TransformFilter<pcl::experimental::VoxelStructT<pcl::PointXYZ>, pcl::PointXYZ>::applyFilter
    13.01%  benchmark_filte  libc-2.31.so                  [.] _int_malloc
     9.38%  benchmark_filte  libc-2.31.so                  [.] _int_free
     6.43%  benchmark_filte  benchmark_filters_voxel_grid  [.] std::_Hashtable<unsigned long, std::pair<unsigned long const, pcl::experimental::Voxel<pcl::PointXYZ> >, std::allocator<std::pair<unsigned long const, pcl::experimental::Voxel<pcl::PointXYZ> > >, std::__detail::_Select1st, std::equal_to<unsigned long>, std::hash<unsigned long>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::_M_emplace<unsigned long const&, bool&>
     5.31%  benchmark_filte  libc-2.31.so                  [.] __mcount_internal
     4.85%  benchmark_filte  benchmark_filters_voxel_grid  [.] pcl::experimental::VoxelStructT<pcl::PointXYZ>::setUp
     4.37%  benchmark_filte  libc-2.31.so                  [.] malloc
     1.45%  benchmark_filte  libpcl_io.so.1.12.0.99        [.] pcl::lzfDecompress
     1.44%  benchmark_filte  libc-2.31.so                  [.] _mcount
     1.27%  benchmark_filte  libc-2.31.so                  [.] cfree@GLIBC_2.2.5
     1.19%  benchmark_filte  libc-2.31.so                  [.] __memset_avx2_erms
     1.11%  benchmark_filte  libpcl_io.so.1.12.0.99        [.] pcl::PCDReader::readBodyBinary
     1.08%  benchmark_filte  libc-2.31.so                  [.] malloc_consolidate
     1.02%  benchmark_filte  libstdc++.so.6.0.28           [.] operator new
```

Reply:
Sorry, I missed the links for the benchmarks for these perfs.